### ISSUE TITLE
API Key and App Key Logged With Failed Requests

### ISSUE THREAD

**Issue by CoryOwens** [DataDog/datadogpy#274]
Currently, if anything goes wrong with a post to Datadog, the API Key and App Key are logged by multiple modules (e.g. `datadog.api`, `urllib3.connectionpool`, `urllib3.util.retry`, etc.) because they are included in the request URL (e.g. `/api/v1/series?api_key=[OBFUSCATED_API_KEY]&application_key=[OBFUSCATED_APP_KEY]`). 

These keys should be sent as request headers, instead. 

```
[   DEBUG] [datadog.api] [pid:3275 ] [    MainThread] Use `requests` based HTTP client.
[   DEBUG] [urllib3.util.retry] [pid:3275 ] [    MainThread] Converted retries value: 3 -> Retry(total=3, connect=None, read=None, redirect=None, status=None)
[   DEBUG] [urllib3.util.retry] [pid:3275 ] [    MainThread] Incremented Retry for (url='/api/v1/series?api_key=[OBFUSCATED_API_KEY]&application_key=[OBFUSCATED_APP_KEY]'): Retry(total=2, connect=None, read=None, redirect=None, status=None)
[ WARNING] [urllib3.connectionpool] [pid:3275 ] [    MainThread] Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<urllib3.connection.VerifiedHTTPSConnection object at 0x7fe8a2e3d5c0>: Failed to establish a new connection: [Errno -2] Name or service not known',)': /api/v1/series?api_key=[OBFUSCATED_API_KEY]&application_key=[OBFUSCATED_APP_KEY]
[   DEBUG] [urllib3.util.retry] [pid:3275 ] [    MainThread] Incremented Retry for (url='/api/v1/series?api_key=[OBFUSCATED_API_KEY]&application_key=[OBFUSCATED_APP_KEY]'): Retry(total=1, connect=None, read=None, redirect=None, status=None)
[ WARNING] [urllib3.connectionpool] [pid:3275 ] [    MainThread] Retrying (Retry(total=1, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<urllib3.connection.VerifiedHTTPSConnection object at 0x7fe8a2e3d630>: Failed to establish a new connection: [Errno -2] Name or service not known',)': /api/v1/series?api_key=[OBFUSCATED_API_KEY]&application_key=[OBFUSCATED_APP_KEY]
[   DEBUG] [urllib3.util.retry] [pid:3275 ] [    MainThread] Incremented Retry for (url='/api/v1/series?api_key=[OBFUSCATED_API_KEY]&application_key=[OBFUSCATED_APP_KEY]'): Retry(total=0, connect=None, read=None, redirect=None, status=None)
[ WARNING] [urllib3.connectionpool] [pid:3275 ] [    MainThread] Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<urllib3.connection.VerifiedHTTPSConnection object at 0x7fe8a2e3d6d8>: Failed to establish a new connection: [Errno -2] Name or service not known',)': /api/v1/series?api_key=[OBFUSCATED_API_KEY]&application_key=[OBFUSCATED_APP_KEY]
[   ERROR] [datadog.api] [pid:3275 ] [    MainThread] Could not request POST https://app.datadoghq.com/api/v1/series: HTTPSConnectionPool(host='app.datadoghq.com', port=443): Max retries exceeded with url: /api/v1/series?api_key=[OBFUSCATED_API_KEY]&application_key=[OBFUSCATED_APP_KEY] (Caused by NewConnectionError('<urllib3.connection.VerifiedHTTPSConnection object at 0x7fe8a2e3d7b8>: Failed to establish a new connection: [Errno -2] Name or service not known',)). Please check the network connection or try again later. If the problem persists, please contact support@datadoghq.com
```

**Comment by chenrui333** [comment]
+1 I have the same issue to submit metrics from AWS lambda

**Comment by vaddipar** [comment]
@CoryOwens  and @chenrui333  since this is a log from a dependency, I suggest you mask logs by have a masking function at your project level. The code for the filter will be something like this.

`import logging


class LoggingFilter(logging.Filter):
    """Filters the Datadog API and Datadog APP keys from the logs."""

    def filter(self, log_record):
        """
        Filters the patterns from the Log record

        :type log_record: logging.LogRecord
        :param log_record: A single logging record that needs filtering (or not)

        :rtype: bool
        :return: True or False
        """

        if isinstance(log_record.args, dict):
            for message in log_record.args:
                log_record.args[message] = self.__mask_data(message)
        else:
            log_record.args = tuple(self.__mask_data(record_arg) for record_arg in log_record.args)

        return True

    def __get_param_value(self, query_string, param_name):
        """
        Given a string if it has the api_key/application_key, get its value

        :type query_string: str
        :param query_string: Query String containing api or app key

        :type param_name: str
        :param param_name: api_key= or application_key=

        :return: value of api or app key
        """

        value = query_string.split(param_name)[-1]
        if ' ' in value:
            value = value.split(' ')[0]
        else:
            value = value.split('&')[0]

        return value

    def __mask_data(self, log_msg):
        """
        Replace the sensitive data in a dictionary

        :type str
        :param log_msg: Log Message

        :rtype: dict
        :return: received dictionary after replacement of sensitive data
        """

        if type(log_msg) != str:
            return log_msg

        if 'api_key' in log_msg:
            api_key_value = self.__get_param_value(log_msg, 'api_key=')
            log_msg = log_msg.replace(api_key_value, '*'*len(api_key_value))

        if 'application_key' in log_msg:
            app_key_value = self.__get_param_value(log_msg, 'application_key=')
            log_msg = log_msg.replace(app_key_value, '*' * len(app_key_value))

        return log_msg`

When you setup logging, you can then add a filter to the logger.

**Comment by CoryOwens** [comment]
@vaddipar - My issue isn't with the library; the library is doing exactly what it is supposed to do. My issue is with how DataDog is using that library. Private keys should be sent as request headers, not as query parameters. Log filtering is a workaround with a noticeable performance cost, not an ideal solution.

**Comment by dil048** [comment]
> ndency, I suggest you mask logs by have a masking function at your project level. The code for the filter will be something like this.

**Comment by katezaps** [comment]
hey 👋 still seeing this now - is there any fix coming forward? we shouldn't be logging API keys at all

**Comment by vaddipar** [comment]
@katezaps  I can fix the client end but how about the API actually reading the api and app key from the headers instead of params?

**Comment by gzussa** [comment]
Thanks everyone for raising this. This is much appreciated :). Indeed, we want to fix this! We raised the problem to the proper team. They actually already knew about it and they will do their best to fix it ASAP. Unfortunately, i cannot provide any ETA at the moment.

**Comment by dabcoder** [comment]
cc @ssc3 as I believe your https://github.com/DataDog/datadogpy/pull/446 PR addresses this issue?

**Comment by ssc3** [comment]
> cc @ssc3 as I believe your #446 PR addresses this issue?

Yup. #446 fixes it. Thanks!

**Comment by zippolyte** [comment]
Actually it doesn't work for all the endpoints at the moment, see https://github.com/DataDog/datadogpy/blob/a944ac604106c5f5dea89c53449d0a05174d5bc7/datadog/api/api_client.py#L246-L259

So this issue is still not fully resolved

**Comment by spiliopoulos** [comment]
This is affecting us as well. Can we revisit if after 3 years its now possible to only send api/app keys through headers and not request parameters?

**Comment by therve** [comment]
Fixed in #754
